src/field/field.rs

```diff
@@ -486,3 +486,212 @@ macro_rules! test_square_root {
+        //     .chain(diff_max)
+        //     .chain(diff_mod)
+        //     .filter(|x| x < &modulus)
+        //     .collect()
+    }
+
+
+    /// Apply the unary functions `op` and `expected_op`
+    /// coordinate-wise to the inputs from `test_inputs(modulus,
+    /// word_bits)` and panic if the two resulting vectors differ.
+    pub fn run_unaryop_test_cases<T, UnaryOp, ExpectedOp>(
+        modulus: &BigUint, word_bits: usize,
+        op: UnaryOp, expected_op: ExpectedOp)
+        -> Result<()>
+    where
+        T: Field,
+        UnaryOp: Fn(T) -> T,
+        ExpectedOp: Fn(BigUint) -> BigUint
+    {
+        let inputs = test_inputs(modulus, word_bits);
+        let expected = inputs.iter().map(|x| expected_op(x.clone()));
+        let output = inputs.iter().map(|x| field_to_biguint(op(biguint_to_field(x.clone()))));
+        // Compare expected outputs with actual outputs
+        assert!(output.zip(expected).all(|(x, y)| x == y),
+                "output differs from expected");
+        Ok(())
+    }
+
+
+    /// Apply the binary functions `op` and `expected_op` to each pair
+    /// in `zip(inputs, rotate_right(inputs, i))` where `inputs` is
+    /// `test_inputs(modulus, word_bits)` and `i` ranges from 0 to
+    /// `inputs.len()`.  Panic if the two functions ever give
+    /// different answers.
+    pub fn run_binaryop_test_cases<T, BinaryOp, ExpectedOp>(
+        modulus: &BigUint, word_bits: usize,
+        op: BinaryOp, expected_op: ExpectedOp)
+        -> Result<()>
+    where
+        T: Field,
+        BinaryOp: Fn(T, T) -> T,
+        ExpectedOp: Fn(BigUint, BigUint) -> BigUint
+    {
+        let inputs = test_inputs(modulus, word_bits);
+
+        for i in 0 .. inputs.len() {
+            // Iterator over inputs rotated right by i places. Since
+            // cycle().skip(i) rotates left by i, we need to rotate by
+            // n_input_elts - i.
+            let shifted_inputs = inputs.iter().cycle().skip(inputs.len() - i);
+            // Calculate pointwise operations
+            let expected = inputs.iter().zip(shifted_inputs.clone()).map(
+                |(x, y)| expected_op(x.clone(), y.clone()));
+            let output = inputs.iter().zip(shifted_inputs).map(
+                |(x, y)| field_to_biguint(op(biguint_to_field(x.clone()),
+                                             biguint_to_field(y.clone()))));
+            // Compare expected outputs with actual outputs
+            assert!(output.zip(expected).all(|(x, y)| x == y),
+                    "output differs from expected at rotation {}", i);
+        }
+        Ok(())
+    }
+}
+
+#[macro_export]
+macro_rules! test_arithmetic {
+    ($field:ty) => {
+        use crate::field_tests;
+        use crate::field_to_biguint;
+        use std::io::Result;
+        use std::ops::{Add,Sub,Mul,Neg,Div};
+        use num::{BigUint, Zero, One};
+
+
+        /// Return the modulus of the type `Fld`.
+        fn field_modulus<Fld>() -> BigUint
```

Comment:
nit: we usually use `F` for Fields. Also could compact it a bit with `F: Field` instead of the where clause.